src/DataStructures/StaticDeque.hpp

```diff
@@ -0,0 +1,520 @@
+    std::rotate(new_position, old_end, this->end());
+    return new_position;
+  } else {
+    // We are about to invalidate position.
+    const auto insert_index = position - this->begin();
+    const auto insert_back_index = this->end() - position;
+    const auto old_size = static_cast<difference_type>(size());
+    for (size_type i = 0; i < n; ++i) {
+      emplace_front(x);
+    }
+    const auto new_position = this->begin() + insert_index;
+    const auto new_position_end = this->end() - insert_back_index;
+    const auto old_begin = this->end() - old_size;
+    std::rotate(this->begin(), old_begin, new_position_end);
+    return new_position;
+  }
+}
+
+template <typename T, size_t Capacity>
+template <typename InputIterator,
+          Requires<static_deque_detail::is_input_iterator<InputIterator>>>
+auto StaticDeque<T, Capacity>::insert(const const_iterator position,
+                                      InputIterator first,
+                                      const InputIterator last) -> iterator {
+  // The iterators could be single-pass, so we can't clear space and
+  // then insert the new elements.  We also can't do repeated single
+  // insertions because that's too inefficient.  Instead, we have to
+  // create the new elements at the end and then rotate them into
+  // place.
+  if (should_operate_on_back(position, position)) {
+    // We are about to invalidate position.
+    const auto insert_index = position - this->begin();
+    const auto old_size = static_cast<difference_type>(size());
+    while (first != last) {
+      emplace_back(*first++);
+    }
+    const auto new_position = this->begin() + insert_index;
+    const auto old_end = this->begin() + old_size;
+    std::rotate(new_position, old_end, this->end());
+    return new_position;
+  } else {
+    // We are about to invalidate position.
+    const auto insert_index = position - this->begin();
+    const auto insert_back_index = this->end() - position;
+    const auto old_size = static_cast<difference_type>(size());
+    while (first != last) {
+      emplace_front(*first++);
+    }
+    const auto new_position = this->begin() + insert_index;
+    const auto new_position_end = this->end() - insert_back_index;
+    const auto old_begin = this->end() - old_size;
+    std::reverse(this->begin(), old_begin);
+    std::rotate(this->begin(), old_begin, new_position_end);
+    return new_position;
+  }
+}
+
+template <typename T, size_t Capacity>
+auto StaticDeque<T, Capacity>::insert(const const_iterator position,
+                                      const std::initializer_list<T> init)
+    -> iterator {
+  return insert(position, init.begin(), init.end());
+}
+
+template <typename T, size_t Capacity>
+void StaticDeque<T, Capacity>::pop_front() {
+  ASSERT(not this->empty(), "Container is empty.");
+  if (start_position_ == Capacity - 1) {
+    start_position_ = 0;
+  } else {
+    ++start_position_;
+  }
+  --size_;
+  entry_address(Capacity - 1)->~T();
```

Comment:
Since explicitly calling destructors looks fishy, maybe add a C++ comment briefly saying it's intentional and why? I worry that someone will think their bug is here in 5 years from now because of some rule they read online 🙃